File: src/ch08/program8_2.py
from program8_1 import func

def main():
    print("分割数を入力してください--->", end="")
    n = int(input())

    heun( 0.0, 1.0, 0.0, 1.0, n, func)


# ホイン法
def heun(x: float, y: float, a: float, b: float, n: int, f):
    h = (b - a) / n

    for _ in range(n):
        k1 = f(x,y)
        k2 = f(x+h, y+h*k1)
        y = y + h/2.0 * ( k1 + k2 )
        x += h
        print("x={:.6f} \t y={:.6f} ".format(x, y))


if __name__ == "__main__":
    main()

File: src/ch08/program8_2.c
#include <stdio.h>

/* 関数の定義 */
double func(double x, double y);
/* ホイン法 */
void heun(double x, double y, double a, double b, int n,
                              double (*f)(double, double) );

int main(void)
{
  int n;

  printf("分割数を入力してください--->");
  scanf("%d",&n);

  heun( 0.0, 1.0, 0.0, 1.0, n, func );

  return 0;
}

/* ホイン法 */
void heun(double x, double y, double a, double b, int n,
                              double (*f)(double, double) )
{
  double k1, k2, h;
  int i;

  h = (b-a)/n;

  for ( i = 0 ; i < n ; i++)
  {
    k1 = f(x,y); k2 = f(x+h, y+h*k1);
    y = y + h/2.0 * ( k1 + k2 );
    x += h;
    printf("x=%f \t y=%f \n", x, y );
  }

}

/* 関数の定義 */
double func(double x, double y)
{
  return( x + y );
}
